network-store-iidm-impl/src/main/java/com/powsybl/network/store/iidm/impl/ConfiguredBusImpl.java

```diff
@@ -80,237 +80,207 @@ public Bus setAngle(double angle) {
 
     @Override
     public boolean isInMainConnectedComponent() {
-        throw new UnsupportedOperationException("TODO");
+        getNetwork().ensureConnectedComponentsUpToDate();
+        Optional<Bus> mergedBus = getMergedBus();
+        return mergedBus.isPresent() && mergedBus.get().isInMainConnectedComponent();
     }
 
     @Override
     public Component getSynchronousComponent() {
-        throw new UnsupportedOperationException("TODO");
+        getNetwork().ensureConnectedComponentsUpToDate();
+        Optional<Bus> mergedBus = getMergedBus();
+        return mergedBus.isPresent() ? mergedBus.get().getSynchronousComponent() : null;
     }
 
     @Override
     public boolean isInMainSynchronousComponent() {
-        throw new UnsupportedOperationException("TODO");
+        getNetwork().ensureConnectedComponentsUpToDate();
+        Optional<Bus> mergedBus = getMergedBus();
+        return mergedBus.isPresent() && mergedBus.get().isInMainSynchronousComponent();
+    }
+
+    @Override
+    public List<Terminal> getConnectedTerminals() {
+        return getTerminals(true);
+    }
+
+    @Override
+    public Stream<Terminal> getConnectedTerminalStream() {
+        return getConnectedTerminals().stream();
+    }
+
+    @Override
+    public List<Terminal> getAllTerminals() {
+        return getTerminals(false);
+    }
+
+    @Override
+    public Stream<Terminal> getAllTerminalsStream() {
+        return getAllTerminals().stream();
+    }
+
+    private List<Terminal> getTerminals(boolean connected) {
+        Predicate<Terminal> pred =
+                connected ?
+                    t -> t.getBusBreakerView().getBus() != null && t.getBusBreakerView().getBus().getId().equals(getId()) :
+                    t -> t.getBusBreakerView().getConnectableBus() != null && t.getBusBreakerView().getConnectableBus().getId().equals(getId());
+        return getVoltageLevel().getConnectableStream()
+                .flatMap(c -> (Stream<Terminal>) c.getTerminals().stream())
+                .filter(t -> t.getVoltageLevel().getId().equals(getVoltageLevel().getId()) && pred.test(t))
+                .collect(Collectors.toList());
     }
 
     @Override
     public int getConnectedTerminalCount() {
-        throw new UnsupportedOperationException("TODO");
+        return getConnectedTerminals().size();
+    }
+
+    private Stream<Connectable> getConnectableStream(ConnectableType type) {
+        return getConnectedTerminals().stream().map(Terminal::getConnectable).filter(c -> c.getType() == type);
     }
 
     @Override
     public Iterable<Line> getLines() {
-        throw new UnsupportedOperationException("TODO");
+        return getLineStream().collect(Collectors.toList());
     }
 
     @Override
     public Stream<Line> getLineStream() {
-        throw new UnsupportedOperationException("TODO");
+        return getConnectableStream(ConnectableType.LINE)
+                .map(c -> index.getLine(c.getId()).orElseThrow(IllegalAccessError::new));
```

Comment:
Wouldn't .map(Line.class::cast) be sufficient here ? (And for later methods of the same type)

Reply:
Yes
Done